### PR TITLE
Support custom http stack

### PR DESCRIPTION
**Pull request by fractalwrench**
Updates the current HttpClient implementation to implement an ErrorReportApiInterface. The Bugsnag class has a new public method in its API to set a custom implementation of this interface. This should allow for custom requirements such as certificate pinning to be achieved.

### CODE REVIEW

**Comment by snmaynard** [comment]
Are we planning on adding docs for how to do this?

File: src/androidTest/java/com/bugsnag/android/ErrorReportApiClientTest.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+            Bugsnag.setErrorReportApiClient(null);
+            fail("ErrorReportApiClient cannot be null");
+        }
+        catch (Exception ignored) {
+        }
+    }
+
+    public void testBugsnagClient() {
+        Bugsnag.init(getContext(), "123");
+        Bugsnag.setErrorReportApiClient(errorReportApiClient);
```
Comment (kattrali) [inline src/androidTest/java/com/bugsnag/android/ErrorReportApiClientTest.java]: Should also test that its `postReport` method is called after a `notify`

**Comment by kattrali** [comment]
@snmaynard I hope so :)

**Comment by kattrali** [comment]
@fractalwrench Looks good, left a couple of comments.

**Comment by fractalwrench** [comment]
@kattrali I've updated this PR according to the review comments, and created a PR on the Docs repo.

**Comment by kattrali** [comment]
🎉

**Pull request merged**